### PR TITLE
Algorithm for finding Peak element added

### PR DESCRIPTION
**Pull request by DevilAJ**
I have added Algorithm for finding Peak element in an array by implementing Divide & Conquer algorithm
in both manner recursively and using while loop.
I also added proper comments and illustration to understand.

### CODE REVIEW

**Comment by atarax665** [comment]
> I have added Algorithm for finding Peak element in an array by implementing Divide & Conquer algorithm
> in both manner recursively and using while loop.
> I also added proper comments and illustration to understand.

Please always open an issue regarding your proposal before making a pull request.

**Comment by DevilAJ** [comment]
@atarax665 
#2274
I did open the issue and it was assigned to me.
You can check.

**Comment by atarax665** [comment]
> @atarax665
> #2274
> I did open the issue and it was assigned to me.
> You can check.

Okay, please do mention the issue related to your PR, you can use the default PR template provided. It allows the maintainers to have a glance on all the necessary details at one place.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+its neighbours. For corner elements, we need to
+consider only one neighbour.
+Since an array always have a maximum value therefore an peak element always be there.
+Eg. :
+Input: array[] = {10, 20, 15, 2, 23, 90, 67}
+Output: 20 or 90
+The element 20 has neighbours 10 and 15, both of
+them are less than 20, similarly 90 has neighbours 23
+and 67.
+A divide-and-conquer algorithm recursively breaks
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Please remove this explanation of what is a divide and conquer algorithm it ain't needed here.

**Comment by DevilAJ** [comment]
Ok sir.
I will make sure to mention issue from now on.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+the same or related type, until these become simple
+enough to be solved directly. The solutions to the
+sub-problems are then combined to give a solution to
+the original problem. **/
+
+#include <bits/stdc++.h>
+using namespace std;
+
+// recursive approach
+//  Complexity : O(log(N)
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Time and space complexity should be added at the end of the program.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+int findPeek(vector<int>v,int l,int r)
+{
+    if(l<r)
+    {
+        int mid = l+(r-l)/2;
+        if(mid == 0 || mid+1 == v.size()) // means we are at corner
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Comments should be above the line of code not beside it.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            {
+                return arr[n-2];
+            }
+        }
+    }
+    return -1;
+}
+
+int main() {
+	ios_base::sync_with_stdio(0);
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Remove ```ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);```.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	cin.tie(0);
+	cout.tie(0);
+	int n;cin>>n;
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Use proper input output statements. It makes your program interactive.

File: C-Plus-Plus/search/Peak_element.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        
+        else 
+        return findPeek(v,l,r-1);
+    }
+    return -1;
+}
+
+// using while loop approach
+//  Complexity : O(log(N)
+int findPeek(vector<int>arr)
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Rename the iterative function as ```find_peak_iter```

File: C-Plus-Plus/search/Peak_element.cpp [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+enough to be solved directly. The solutions to the
+sub-problems are then combined to give a solution to
+the original problem. **/
+
+#include <bits/stdc++.h>
+using namespace std;
+
+// recursive approach
+//  Complexity : O(log(N)
+int findPeek(vector<int>v,int l,int r)
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Rename the iterative function as ```find_peak_recursive```

File: C-Plus-Plus/search/Peak_element.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+int main() {
+	ios_base::sync_with_stdio(0);
+	cin.tie(0);
+	cout.tie(0);
+	int n;cin>>n;
+	vector<int> v(n);
+	for(int i=0;i<n;i++)
+	cin>>v[i];
+	int i = findPeek(v,0,n-1);
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: Use proper variable names and as you have added 2 different approaches then format your output accordingly.
ex : 
```
Peak Element (Recursive approach) : 
Peak Element (Iterative approach) :
```

File: C-Plus-Plus/search/Peak_element.cpp [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+	cout << i <<"\n";
+	return 0;
+}
+
+/** ILLUSTRATION
+Input : a[] = { 10, 20, 15, 2, 23, 90, 6 }
```
Comment (atarax665) [inline C-Plus-Plus/search/Peak_element.cpp]: You just need to add Sample Input/Output and complexity at the end. You can explain your approach briefly either using inline comments or adding a docstring before your functions.

**Review by atarax665 — CHANGES_REQUESTED** [review]
Please make appropriate changes to the corresponding readme and change the file location from ```C-Plus-Plus/search to C-Plus-Plus/cp```

**Comment by DevilAJ** [comment]
Ok sir and thanx for your time.
I will make required changes.

**Comment by DevilAJ** [comment]
I had done the required changes.
Do let me know for any further updation.

**Comment by DevilAJ** [comment]
@atarax665 
Please check the changes 
And let me know for any changes.

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+                return arr[n-2];
+            }
+        }
+    }
+    return -1;
+}
+
+int main() {
+	int n;
+	cin>>n;
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Prompt user to ```Enter the number of elements in the array :``` and ```Enter the elements of the array :```

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	vector<int> v(n);
+	for(int i=0;i<n;i++)
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Add opening and closing braces of the for loop.

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+	
+	int peak_iter = find_peak_iter(v);
+	cout << "Peak element ( Iterative approach ) : "<< peak_iter <<"\n";
+	
+	return 0;
+}
+
+/** 
+Eg. :
+Input: array[] = {10, 20, 15, 2, 23, 90, 67}
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Your sample input should replicate the input you will give when the program is run in console or any IDE.

File: C-Plus-Plus/cp/README.md [hunk 13]
```diff
@@ -1 +1,6 @@
-a
+Peak Element
```
Comment (atarax665) [inline C-Plus-Plus/cp/README.md]: Revert this. You don't need to update this sub-folder README. You just need to add hyperlink of your program in the README file of C-Plus-Plus folder. Refer to a recently merged C++ PR to get a better idea of this.

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+Output: 
+Peak element ( Recursive approach ) : 90
+Peak element ( Iterative approach ) : 20
+
+Time Complexity : O(log(N))
+Space Complexity : O(N)
+Approach :If the middle element is not the peak element, then check if the element on the right side is greater than the middle element 
+then there is always a peak element on the right side. 
+If the element on the left side is greater than the middle element then there is always a peak element on the left side.
+**/
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Add a blank line at the EOF.

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+Eg. :
+Input: array[] = {10, 20, 15, 2, 23, 90, 67}
+
+Output: 
+Peak element ( Recursive approach ) : 90
+Peak element ( Iterative approach ) : 20
+
+Time Complexity : O(log(N))
+Space Complexity : O(N)
+Approach :If the middle element is not the peak element, then check if the element on the right side is greater than the middle element
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Shift this approach paragraph after the description part at the top.

**Review by atarax665 — CHANGES_REQUESTED** [review]
There are just some minor fixes and then this PR will be good to merge. Also format your code using [this](https://techiedelight.com/tools/clike).

**Comment by DevilAJ** [comment]
@atarax665 
I have done the changes you can check.

**Comment by DevilAJ** [comment]
@atarax665 
Please check it.
I think it will show conflicts after some time.

**Comment by DevilAJ** [comment]
@atarax665 
@HarshCasper 
@bhav09 
@gat786 
Please look into this pr it still pending....

File: C-Plus-Plus/cp/Peak_element.cpp [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+	int peak_iter = find_peak_iter(v);
+	cout << "Peak element (Iterative approach) : " << peak_iter << "\n";
+
+	return 0;
+}
+
+/**
+Eg. :
+Input: 
+7
```
Comment (atarax665) [inline C-Plus-Plus/cp/Peak_element.cpp]: Update the sample input according to the updated code.

File: C-Plus-Plus/README.md [hunk 17]
```diff
@@ -52,8 +52,11 @@
 - [Target Sum triplets](cp/target_sum_triplets.cpp)
 - [Two Pointers Method](cp/two_pointers_method.cpp)
 - [Z Algorithm](cp/zalgorithm.cpp)
+<<<<<<< HEAD
```
Comment (atarax665) [inline C-Plus-Plus/README.md]: Remove this line.

File: C-Plus-Plus/README.md [hunk 18]
```diff
@@ -52,8 +52,11 @@
...
+- [Peak element](cp/Peak_element.cpp)
 - [First occurence of 1](cp/first1_binsearch.cpp)
 - [Remove Minimum Parantheses to make Valid Parantheses](cp/removeMinParantheses.cpp)
+>>>>>>> c8ecee37a2708f6788e13c1e0080f40635dd0702
```
Comment (atarax665) [inline C-Plus-Plus/README.md]: Remove this line.

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by DevilAJ** [comment]
@atarax665 
Please look now..

**Comment by DevilAJ** [comment]
@atarax665 
My issue had been closed and PR is still unmerged please look into it.

File: C-Plus-Plus/README.md [hunk 19]
```diff
@@ -58,11 +58,18 @@
 - [Toggle Kth Bit From Right](cp/ToggleKthBitFromRight.cpp)
 - [Two Pointers Method](cp/two_pointers_method.cpp)
 - [Z Algorithm](cp/zalgorithm.cpp)
-<<<<<<< HEAD
-- [Peak element](cp/Peak_element.cpp)
 - [First occurence of 1](cp/first1_binsearch.cpp)
 - [Remove Minimum Parantheses to make Valid Parantheses](cp/removeMinParantheses.cpp)
->>>>>>> c8ecee37a2708f6788e13c1e0080f40635dd0702
+- [Moore's Voting Algorithm to find the majority element](cp/moore_voting.cpp)
```
Comment (atarax665) [inline C-Plus-Plus/README.md]: Do not make unnecessary changes in the README. Just add hyper link of your program under the correct heading.

**Pull request merged**